pkg/network/ebpf/c/protocols/classification/protocol-classification.h

```diff
@@ -87,119 +140,88 @@ __maybe_unused static __always_inline void protocol_classifier_entrypoint(struct
-    //
-    // The connection is TLS encrypted, thus we cannot classify the protocol using socket filter.
-    if (is_tls_connection_cached(&skb_tup)) {
+    usm_context_t *usm_ctx = usm_context_init(skb, &skb_tup, &skb_info);
+    if (!usm_ctx) {
         return;
     }
 
-    protocol_t *cur_fragment_protocol_ptr = bpf_map_lookup_elem(&connection_protocol, &skb_tup);
-    if (cur_fragment_protocol_ptr) {
+    protocol_stack_t *protocol_stack = get_protocol_stack(skb, usm_ctx);
+    if (!protocol_stack) {
         return;
     }
 
-    // Get the buffer the fragment will be read into from a per-cpu array map.
-    // This will avoid doing unaligned stack access while parsing the protocols,
-    // which is forbidden and will make the verifier fail.
-    const u32 key = 0;
-    char *request_fragment = bpf_map_lookup_elem(&classification_buf, &key);
-    if (request_fragment == NULL) {
-        log_debug("could not get classification buffer from map");
+    if (is_fully_classified(protocol_stack)) {
         return;
     }
 
-    bpf_memset(request_fragment, 0, sizeof(request_fragment));
-    read_into_buffer_for_classification((char *)request_fragment, skb, skb_info.data_off);
-    const size_t payload_length = skb->len - skb_info.data_off;
-    const size_t final_fragment_size = payload_length < CLASSIFICATION_MAX_BUFFER ? payload_length : CLASSIFICATION_MAX_BUFFER;
-
-    // In the context of socket filter, we can classify the protocol if it is plain text,
-    // so if the protocol is encrypted, then we have to rely on our uprobes to classify correctly the protocol.
-    if (is_tls(request_fragment, final_fragment_size)) {
-        const bool t = true;
-        bpf_map_update_with_telemetry(tls_connection, &skb_tup, &t, BPF_ANY);
-        flip_tuple(&skb_tup);
-        bpf_map_update_with_telemetry(tls_connection, &skb_tup, &t, BPF_ANY);
+    const char *buffer = &(usm_ctx->buffer.data[0]);
+    // TLS classification
+    if (is_tls(buffer, usm_ctx->buffer.size)) {
+        set_protocol_stack(skb, protocol_stack, PROTOCOL_TLS);
+        mark_as_fully_classified(protocol_stack);
         return;
     }
 
-    protocol_t cur_fragment_protocol = classify_applayer_protocols(request_fragment, final_fragment_size);
-    // If there has been a change in the classification, save the new protocol.
-    if (cur_fragment_protocol != PROTOCOL_UNKNOWN) {
-        save_protocol(&skb_tup, cur_fragment_protocol);
-        return;
+    // If application-layer is known we don't bother to check for HTTP protocols and skip to the next layers
+    if (protocol_layer_known(protocol_stack, LAYER_APPLICATION)) {
+        goto next_program;
     }
 
-    bpf_tail_call_compat(skb, &classification_progs, CLASSIFICATION_QUEUES_PROG);
+    protocol_t cur_fragment_protocol = classify_applayer_protocols(buffer, usm_ctx->buffer.size);
+    set_protocol_stack(skb, protocol_stack, cur_fragment_protocol);
+    if (cur_fragment_protocol) {
```

Comment:
if the protocol is known, should we continue to the next classification program?

Reply:
in theory we should, but currently there are no other protocols that would be classified in addition to HTTP.
Removing the
```
mark_as_fully_classified(protocol_stack);
return;
```
would (currently) result in the same behavior, just would go through a slightly longer code path